### DISCUSSION TITLE
Confusion with custom macros and combos

### DISCUSSION THREAD

**Discussion by Kirsturino** [manna-harbour/miryoku#233] [Q&A] [answered]
I'm setting up my ZMK layout for a 5-column corne and got a custom layer and some simple key swaps up and running. Great! I'd also like to add my own macros and combos to my layout, but am unsure how to do so. I've read the readme and customization document multiple times and delved into the configs to try and figure this out, but no success so far. Even if I try to copy & paste existing macros with minimal changes, all my builds seem to fail.

Am I missing something? Are custom macros and/or combos supported at all? Could anyone provide a simple example macro/combo they've set up so I could see how to do so?

**Comment by manna-harbour** [comment, marked as answer]
[`custom_config.h`](https://github.com/manna-harbour/miryoku/discussions/85) is intended just for `cpp` macros (e.g. `#define`), and in future there will be an analogous `custom.keymap` or similar for device tree. But given the way ZMK keymaps work, you can add keymap devicetree to `custom_config.h` and it'll be equivalent, so just do that for now. Note that you'll need to specify the node from the root.

So for macros, you can do that and then reference them in your custom layer etc. There's an example macro here https://github.com/manna-harbour/miryoku_zmk/blob/0865642f2e4f534a68667113e80a784435942c3f/miryoku/miryoku_kludge_tapdelay.h#L6, using this `cpp` macro https://github.com/manna-harbour/miryoku_zmk/blob/0865642f2e4f534a68667113e80a784435942c3f/miryoku/miryoku.h#L11.  You can use that if it's suitable, or just do it directly in devicetree without any `cpp` macros.

For combos, there's an example here https://github.com/manna-harbour/miryoku_zmk/blob/master/miryoku/miryoku_kludge_thumbcombos.dtsi and https://github.com/manna-harbour/miryoku_zmk/blob/0865642f2e4f534a68667113e80a784435942c3f/miryoku/mapping/42/corne.h#L29. Again you can use a similar `cpp` macro or do it directly in bare devicetree. You can define the key positions macros within an `#ifdef` per mapping as described in the section on custom mappings in the customisation docs.

If you're running into build errors, you might like to try it directly in `corne.keymap` first and without any `cpp` macros, and building locally.

**Comment by Kirsturino** [comment]
Thanks for the response! I'm still a beginner with ZMK so it's taking me a while to understand all this. Things like specifying the node from the root I don't understand yet. This gives me a direction to head in, though, so I'll give it a shot when I get the chance!
